File: Ejercicios/Kruskal/Haskell/Kruskal.hs
module Kruskal(kruskal, kruskals) where

import Data.Maybe(fromJust)
import qualified DataStructures.Dictionary.AVLDictionary as D
import qualified DataStructures.PriorityQueue.LinearPriorityQueue as Q
import DataStructures.Graph.DictionaryWeightedGraph

kruskal :: (Ord a, Ord w) => WeightedGraph a w -> [WeightedEdge a w]
kruskal graph = kruskal' pq dict []
  where
      pq   = foldr (Q.enqueue) (Q.empty) (edges graph)
      dict = foldr (\v d -> D.insert v v d) (D.empty) (vertices graph)
      kruskal' q d f | Q.isEmpty q = f 
                     | r1 == r2    = kruskal' (Q.dequeue q) d f
                     | otherwise   = kruskal' (Q.dequeue q) (D.insert r2 v1 d) (e:f)
                       where
                           r1 = representante v1 dict
                           r2 = representante v2 dict
                           e@(WE v1 w v2) = Q.first q

representante :: (Ord a) => a -> D.Dictionary a a -> a
representante v dict | v == w    = v
                     | otherwise = representante w dict
                        where
                            w    = fromJust(D.valueOf v dict)

-- Solo para evaluación continua / only for part time students
kruskals :: (Ord a, Ord w) => WeightedGraph a w -> [[WeightedEdge a w]]
kruskals = undefined

File: Ejercicios/Kruskal/Java/Kruskal.java

import dataStructures.graph.WeightedGraph;
import dataStructures.graph.WeightedGraph.WeightedEdge;

import dataStructures.dictionary.Dictionary;
import dataStructures.dictionary.HashDictionary;
import dataStructures.priorityQueue.PriorityQueue;
import dataStructures.priorityQueue.LinkedPriorityQueue;
import dataStructures.set.Set;
import dataStructures.set.HashSet;

public class Kruskal {

	public static <V,W> Set<WeightedEdge<V,W>> kruskal(WeightedGraph<V,W> g) {

		Set<WeightedEdge<V,W>> forest = new HashSet<>();

		PriorityQueue<WeightedEdge<V,W>> PQ = new LinkedPriorityQueue();
		for(WeightedEdge<V,W> edge : g.edges()) PQ.enqueue(edge);

		Dictionary<V,V> dict = new HashDictionary<>();
		for(V vertex : g.vertices()) dict.insert(vertex,vertex);

		while(!PQ.isEmpty()){
			WeightedEdge<V,W> e = PQ.first();
			PQ.dequeue();

			V r1 = representante(e.source(),dict);
			V r2 = representante(e.destination(),dict);

			if(!r1.equals(r2)){
				dict.insert(r2,e.source());
				forest.insert(e);
			}
		}
		return forest;
	}

	private static <V> V representante(V src, Dictionary<V,V> dict){
		V v = src;
		V r = dict.valueOf(v);

		while(!v.equals(r)){
			v = r;
			r = dict.valueOf(v);
		}

		return r;
	}

	// Sólo para evaluación continua / only for part time students
	public static <V,W> Set<Set<WeightedEdge<V,W>>> kruskals(WeightedGraph<V,W> g) {

		// COMPLETAR
		
		return null;
	}
}
